File: src/main/python/com/hazza/LeetCodeByPython/SegmentTwo/No73_Set_Matrix_Zeroes.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019-07-04
# https://leetcode.com/problems/set-matrix-zeroes/
from typing import List


class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        rowLen, colLen = len(matrix), len(matrix[0])
        isFirstColZero = False

        for i in range(rowLen):
            if not matrix[i][0]:
                isFirstColZero = True
            for j in range(1, colLen):
                if not matrix[i][j]:
                    matrix[i][0] = matrix[0][j] = 0

        for i in range(0, rowLen)[::-1]:
            for j in range(1, colLen)[::-1]:
                if not matrix[i][0] or not matrix[0][j]:
                    matrix[i][j] = 0
            if isFirstColZero:
                matrix[i][0] = 0


if __name__ == '__main__':
    solution = Solution()

    input = [
        [1, 1, 1],
        [1, 0, 1],
        [1, 1, 1]
    ]
    output = [
        [1, 0, 1],
        [0, 0, 0],
        [1, 0, 1]
    ]
    solution.setZeroes(input)
    assert input == output

    input = [
        [0, 1, 2, 0],
        [3, 4, 5, 2],
        [1, 3, 1, 5]
    ]
    output = [
        [0, 0, 0, 0],
        [0, 4, 5, 0],
        [0, 3, 1, 0]
    ]
    solution.setZeroes(input)
    assert input == output

File: src/main/java/com/hazza/leetcode/SegmentTwo/No73SetMatrixZeroes.java
package com.hazza.leetcode.SegmentTwo;

/**
 * Created with IntelliJ IDEA.
 * Description: https://leetcode.com/problems/set-matrix-zeroes/description/
 * Author: HazzaCheng
 * Contact: hazzacheng@gmail.com
 * Date: 2017-08-26
 * Time: 9:37 AM
 */
public class No73SetMatrixZeroes {
    public void setZeroes(int[][] matrix) {
        int rowLen = matrix.length, colLen = matrix[0].length;
        boolean firstColIs0 = false;
        for (int i = 0; i < rowLen; i++) {
            if (matrix[i][0] == 0) firstColIs0 = true;
            for (int j = 1; j < colLen; j++)
                if (matrix[i][j] == 0)
                    matrix[i][0] = matrix[0][j] = 0;
        }
        for (int i = rowLen - 1; i >= 0; i--) {
            for (int j = colLen - 1; j >= 1; j--) {
                if (matrix[i][0] == 0 || matrix[0][j] == 0) matrix[i][j] = 0;
            }
            if (firstColIs0) matrix[i][0] = 0;
        }
    }
}
